File: sdk/typescript/src/patch-langchain.ts
/**
 * LangChain.js (>= 0.1) 全局 patch — 一行启用 LangChain 全自动 trace.
 *
 * 用法:
 * ```ts
 * import { patchLangChainJS } from '@iaoep/sdk';
 * patchLangChainJS();
 *
 * // 之后所有 ChatOpenAI / Tool 调用都自动 trace
 * const llm = new ChatOpenAI({ model: 'qwen3-turbo' });
 * const result = await llm.invoke('Hello');
 * ```
 */

import { SpanStatusCode } from '@opentelemetry/api';
import { getTracer } from './config.js';

export function patchLangChainJS(): void {
  patchChatModels();
  patchTools();
}

function patchChatModels(): void {
  // 尝试加载 @langchain/core
  let core: typeof import('@langchain/core') | undefined;
  try {
    // dynamic require to make it optional
    // @ts-expect-error - optional dep
    core = require('@langchain/core');
  } catch {
    console.info('[iaoep-sdk] @langchain/core not installed, skipping ChatModel patch');
    return;
  }
  if (!core) return;

  const tracer = getTracer();
  const BaseChatModel = (core as Record<string, unknown>).BaseChatModel as
    | { prototype: Record<string, unknown> }
    | undefined;
  if (!BaseChatModel) return;

  const proto = BaseChatModel.prototype as Record<string, unknown>;
  const origInvoke = proto.invoke as (...a: unknown[]) => Promise<unknown>;
  if (origInvoke) {
    proto.invoke = async function patchedInvoke(this: unknown, ...rest: unknown[]) {
      const modelName = (this as Record<string, unknown>).modelName
        ?? (this as Record<string, unknown>).model
        ?? 'unknown';
      const span = tracer.startSpan('llm.call');
      span.setAttribute('gen_ai.system', 'langchain');
      span.setAttribute('gen_ai.request.model', String(modelName));
      try {
        const result = await origInvoke.apply(this, rest);
        extractTokens(span, result);
        span.setStatus({ code: SpanStatusCode.OK });
        return result;
      } catch (e) {
        span.setStatus({
          code: SpanStatusCode.ERROR,
          message: e instanceof Error ? e.message : String(e),
        });
        throw e;
      } finally {
        span.end();
      }
    };
  }
}

function patchTools(): void {
  let core: typeof import('@langchain/core') | undefined;
  try {
    // @ts-expect-error - optional dep
    core = require('@langchain/core');
  } catch {
    return;
  }
  if (!core) return;

  const tracer = getTracer();
  const StructuredTool = (core as Record<string, unknown>).StructuredTool as
    | { prototype: Record<string, unknown> }
    | undefined;
  if (!StructuredTool) return;

  const proto = StructuredTool.prototype as Record<string, unknown>;
  const origCall = proto._call as (...a: unknown[]) => Promise<unknown>;
  if (origCall) {
    proto._call = async function patchedCall(this: unknown, ...rest: unknown[]) {
      const name = (this as Record<string, unknown>).name ?? 'unknown_tool';
      const span = tracer.startSpan('tool.execute');
      span.setAttribute('tool.name', String(name));
      try {
        const result = await origCall.apply(this, rest);
        span.setStatus({ code: SpanStatusCode.OK });
        return result;
      } catch (e) {
        span.setStatus({
          code: SpanStatusCode.ERROR,
          message: e instanceof Error ? e.message : String(e),
        });
        throw e;
      } finally {
        span.end();
      }
    };
  }
}

function extractTokens(span: { setAttribute: (k: string, v: unknown) => void }, result: unknown): void {
  if (!result || typeof result !== 'object') return;
  const r = result as Record<string, unknown>;

  // LangChain AIMessage: result.usage_metadata.input_tokens / output_tokens
  if (r.usage_metadata && typeof r.usage_metadata === 'object') {
    const usage = r.usage_metadata as Record<string, unknown>;
    if (usage.input_tokens !== undefined) {
      span.setAttribute('gen_ai.usage.input_tokens', Number(usage.input_tokens));
    }
    if (usage.output_tokens !== undefined) {
      span.setAttribute('gen_ai.usage.output_tokens', Number(usage.output_tokens));
    }
  }

  // 备选: result.response_metadata.token_usage
  if (r.response_metadata && typeof r.response_metadata === 'object') {
    const rm = r.response_metadata as Record<string, unknown>;
    const usage = rm.token_usage as Record<string, unknown> | undefined;
    if (usage) {
      if (usage.prompt_tokens !== undefined && r.usage_metadata === undefined) {
        span.setAttribute('gen_ai.usage.input_tokens', Number(usage.prompt_tokens));
      }
      if (usage.completion_tokens !== undefined && r.usage_metadata === undefined) {
        span.setAttribute('gen_ai.usage.output_tokens', Number(usage.completion_tokens));
      }
    }
  }
}

File: sdk/python/src/iaoep_sdk/patch_langchain.py
"""LangChain 全局 patch — 一行启用 LangChain 全自动 trace.

用法:
    from iaoep_sdk import patch_langchain
    patch_langchain()  # 全局启用, 之后所有 LangChain 调用都自动 trace

支持 LangChain >= 0.1 (langchain-core).
"""

from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)


def patch_langchain() -> None:
    """Patch LangChain BaseChatModel.invoke / BaseTool.run 自动产生 Span.

    安全调用: 若 langchain-core 未安装, 则跳过 patch (不抛异常).
    """
    try:
        from langchain_core.language_models.chat_models import BaseChatModel
        from langchain_core.tools import BaseTool
    except ImportError:
        logger.info("langchain-core not installed, skipping patch")
        return

    _patch_chat_model(BaseChatModel)
    _patch_tool(BaseTool)
    logger.info("LangChain patched successfully")


def _patch_chat_model(BaseChatModel: Any) -> None:
    """Patch BaseChatModel.invoke / ainvoke / stream."""
    from iaoep_sdk.config import get_tracer

    tracer = get_tracer()
    original_invoke = BaseChatModel.invoke
    original_ainvoke = BaseChatModel.ainvoke
    original_stream = BaseChatModel.stream

    def patched_invoke(self, *args, **kwargs):
        model = getattr(self, "model_name", None) or getattr(self, "model", "unknown")
        with tracer.start_as_current_span("llm.call") as span:
            span.set_attribute("gen_ai.system", "langchain")
            span.set_attribute("gen_ai.request.model", str(model))
            try:
                result = original_invoke(self, *args, **kwargs)
                _extract_tokens(span, result)
                return result
            except Exception as e:
                from opentelemetry.trace import Status, StatusCode
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def patched_ainvoke(self, *args, **kwargs):
        model = getattr(self, "model_name", None) or getattr(self, "model", "unknown")
        with tracer.start_as_current_span("llm.call") as span:
            span.set_attribute("gen_ai.system", "langchain")
            span.set_attribute("gen_ai.request.model", str(model))
            try:
                result = await original_ainvoke(self, *args, **kwargs)
                _extract_tokens(span, result)
                return result
            except Exception as e:
                from opentelemetry.trace import Status, StatusCode
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    def patched_stream(self, *args, **kwargs):
        model = getattr(self, "model_name", None) or getattr(self, "model", "unknown")
        with tracer.start_as_current_span("llm.call") as span:
            span.set_attribute("gen_ai.system", "langchain")
            span.set_attribute("gen_ai.request.model", str(model))
            try:
                yield from original_stream(self, *args, **kwargs)
            except Exception as e:
                from opentelemetry.trace import Status, StatusCode
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    BaseChatModel.invoke = patched_invoke
    BaseChatModel.ainvoke = patched_ainvoke
    BaseChatModel.stream = patched_stream


def _patch_tool(BaseTool: Any) -> None:
    """Patch BaseTool.run / arun."""
    from iaoep_sdk.config import get_tracer

    tracer = get_tracer()
    original_run = BaseTool.run
    original_arun = BaseTool.arun

    def patched_run(self, *args, **kwargs):
        name = getattr(self, "name", "unknown_tool")
        with tracer.start_as_current_span("tool.execute") as span:
            span.set_attribute("tool.name", str(name))
            try:
                return original_run(self, *args, **kwargs)
            except Exception as e:
                from opentelemetry.trace import Status, StatusCode
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    async def patched_arun(self, *args, **kwargs):
        name = getattr(self, "name", "unknown_tool")
        with tracer.start_as_current_span("tool.execute") as span:
            span.set_attribute("tool.name", str(name))
            try:
                return await original_arun(self, *args, **kwargs)
            except Exception as e:
                from opentelemetry.trace import Status, StatusCode
                span.set_status(Status(StatusCode.ERROR, str(e)))
                raise

    BaseTool.run = patched_run
    BaseTool.arun = patched_arun


def _extract_tokens(span: Any, result: Any) -> None:
    """从 LangChain AIMessage 提取 token 数."""
    usage = None
    # AIMessage 多种结构: result.usage_metadata, result.response_metadata, etc.
    if hasattr(result, "usage_metadata") and result.usage_metadata:
        usage = result.usage_metadata
    elif hasattr(result, "response_metadata"):
        rm = result.response_metadata or {}
        if "token_usage" in rm:
            usage = rm["token_usage"]

    if usage:
        in_tok = usage.get("input_tokens") or usage.get("prompt_tokens")
        out_tok = usage.get("output_tokens") or usage.get("completion_tokens")
        if in_tok:
            span.set_attribute("gen_ai.usage.input_tokens", int(in_tok))
        if out_tok:
            span.set_attribute("gen_ai.usage.output_tokens", int(out_tok))
